File: cpp/lib/testcontainers/core/Mount.cpp
#include <rust/cxx.h>
#include <rust_tc_bridge/lib.h>

#include "testcontainers/core/Mount.hpp"

#include "details/BoxHelper.hpp"

namespace testcontainers {

Mount::Mount(RsMount *mount) noexcept
    : rimpl_(mount, [](RsMount *m) { ::rs_mount_destroy(details::box_from_raw(m)); }) {}

Mount::Mount(Mount &&other) noexcept = default;

Mount &Mount::operator=(Mount &&other) noexcept = default;

Mount::~Mount() noexcept = default;

Mount Mount::BindMount(std::string_view host_path, std::string_view container_path) noexcept {
  return Mount(::rs_mount_bind_mount(details::into_string(host_path), details::into_string(container_path)).into_raw());
}

Mount Mount::VolumeMount(std::string_view name, std::string_view container_path) noexcept {
  return Mount(::rs_mount_volume_mount(details::into_string(name), details::into_string(container_path)).into_raw());
}

Mount Mount::TmpfsMount(std::string_view container_path) noexcept {
  return Mount(::rs_mount_tmpfs_mount(details::into_string(container_path)).into_raw());
}

Mount Mount::with_read_only() noexcept {
  return Mount(::rs_mount_with_read_only(details::into_box(rimpl_)).into_raw());
}

Mount Mount::with_read_write() noexcept {
  return Mount(::rs_mount_with_read_write(details::into_box(rimpl_)).into_raw());
}

Mount Mount::with_size_bytes(std::int64_t size) noexcept {
  return Mount(::rs_mount_with_size_bytes(details::into_box(rimpl_), size).into_raw());
}

Mount Mount::with_mode(std::int64_t mode) noexcept {
  return Mount(::rs_mount_with_mode(details::into_box(rimpl_), mode).into_raw());
}

bool Mount::is_valid() const noexcept { return static_cast<bool>(rimpl_); }

} // namespace testcontainers

File: rust/src/core/mount.rs
use testcontainers::core::{AccessMode, Mount};

pub struct RsMount {
    mount: Mount,
}

pub fn rs_mount_bind_mount(host_path: String, container_path: String) -> Box<RsMount> {
    Box::new(RsMount::new(Mount::bind_mount(host_path, container_path)))
}

pub fn rs_mount_volume_mount(name: String, container_path: String) -> Box<RsMount> {
    Box::new(RsMount::new(Mount::volume_mount(name, container_path)))
}

pub fn rs_mount_tmpfs_mount(container_path: String) -> Box<RsMount> {
    Box::new(RsMount::new(Mount::tmpfs_mount(container_path)))
}

pub fn rs_mount_with_read_only(mount: Box<RsMount>) -> Box<RsMount> {
    Box::new(RsMount::new(mount.mount.with_access_mode(AccessMode::ReadOnly)))
}

pub fn rs_mount_with_read_write(mount: Box<RsMount>) -> Box<RsMount> {
    Box::new(RsMount::new(mount.mount.with_access_mode(AccessMode::ReadWrite)))
}

pub fn rs_mount_with_size_bytes(mount: Box<RsMount>, size: i64) -> Box<RsMount> {
    Box::new(RsMount::new(mount.mount.with_size_bytes(size)))
}

pub fn rs_mount_with_mode(mount: Box<RsMount>, mode: i64) -> Box<RsMount> {
    Box::new(RsMount::new(mount.mount.with_mode(mode)))
}

pub fn rs_mount_destroy(mount: Box<RsMount>) {
    drop(mount);
}

impl RsMount {
    pub fn new(mount: Mount) -> Self {
        Self { mount }
    }
}

impl Into<Mount> for RsMount {
    fn into(self) -> Mount {
        self.mount
    }
}
